File: 2022/09.py
import sys
from typing import DefaultDict

diff = {
    'D': (0, 1),
    'U': (0, -1),
    'L': (-1, 0),
    'R': (1, 0),
}

def solve(lines, length):
    rope = [(0, 0)] * length
    tailpositions = {(0, 0)}

    for direction, steps in lines:
        for _ in range(steps):
            head0, head1 = rope[0]
            diff0, diff1 = diff[direction]
            rope[0] = head0+diff0, head1+diff1

            for knot in range(length-1):
                head0, head1 = rope[knot]
                tail0, tail1 = rope[knot+1]

                if tail0 == head0 and tail1 < head1 - 1:
                    tail1 = head1 - 1
                elif tail0 == head0 and tail1 > head1 + 1:
                    tail1 = head1 + 1
                elif tail1 == head1 and tail0 < head0 - 1:
                    tail0 = head0 - 1
                elif tail1 == head1 and tail0 > head0 + 1:
                    tail0 = head0 + 1
                elif tail0 != head0 and tail1 != head1:
                    diff_lr = head0 - tail0
                    diff_ud = head1 - tail1

                    if abs(diff_lr) > 1 or abs(diff_ud) > 1:
                        tail0 += 1 if diff_lr > 0 else -1
                        tail1 += 1 if diff_ud > 0 else -1

                rope[knot+1] = tail0, tail1

            tailpositions.add(rope[-1])

    return len(tailpositions)


def main():
    lines = []
    for line in sys.stdin:
        direction, steps = line.split(' ')
        lines.append((direction, int(steps)))

    print('part_one', solve(lines, 2))

    print('part_two', solve(lines, 10))


main()

File: 2022/09.js
const readline = require('readline')
const rl = readline.createInterface({ input: process.stdin })

const lines = []

rl.on('line', data => {
  const [direction, steps] = data.split(' ')

  lines.push({ direction, steps: parseInt(steps) })
})

rl.on('close', () => {
  console.log('partOne', partOne(lines))
  console.log('partTwo', partTwo(lines))
})

function partOne (lines) {
  return solve(lines, 2)
}
function partTwo (lines) {
  return solve(lines, 10)
}
function solve (lines, length) {
  const rope = Array.from({ length }, () => [0, 0])

  const tailpositions = new Set()

  for (const { direction, steps } of lines) {
    for (let i = 0; i < steps; i += 1) {
      let head = rope[0]
      let tail = rope[1]
      switch (direction) {
        case 'U':
          head[1] -= 1
          break
        case 'D':
          head[1] += 1
          break
        case 'L':
          head[0] -= 1
          break
        case 'R':
          head[0] += 1
          break
        default:
        // noop
      }

      for (let knot = 0; knot < rope.length - 1; knot += 1) {
        head = rope[knot]
        tail = rope[knot + 1]

        if (tail[0] === head[0] && tail[1] < head[1] - 1) {
          tail[1] = head[1] - 1
        } else

        if (tail[0] === head[0] && tail[1] > head[1] + 1) {
          tail[1] = head[1] + 1
        } else

        if (tail[1] === head[1] && tail[0] < head[0] - 1) {
          tail[0] = head[0] - 1
        } else

        if (tail[1] === head[1] && tail[0] > head[0] + 1) {
          tail[0] = head[0] + 1
        } else

        if (tail[0] !== head[0] && tail[1] !== head[1]) {
          const lr = head[0] - tail[0]
          const ud = head[1] - tail[1]

          if (Math.abs(lr) > 1 || Math.abs(ud) > 1) {
            tail[0] += (lr > 0 ? 1 : -1)
            tail[1] += (ud > 0 ? 1 : -1)
          }
        }
      }

      tailpositions.add(tail.join(','))
    }
  }

  return tailpositions.size // 2527 is too high
}
